compiler/backend/src/org/jetbrains/kotlin/codegen/coroutines/coroutines-codegen.md

```diff
@@ -0,0 +1,3263 @@
+exception.
+2. It calls `completion` continuation's `resumeWith` when the coroutine completes so that execution returns to the caller.
+3. It catches exceptions and resumes the `completion` with it, wrapped in `Result`, thus propagating the exception to the caller.
+
+Additionally, `resumeWith` calls `releaseIntercepted` upon coroutine's completion to clear the interceptor up.
+
+The compiler generates `create` overrides for suspending lambdas with zero or one parameter, since `createCoroutineUnintercepted` calls
+them.
+
+The rest (`callerFrame` and `getStackTraceElement`) come from the `CoroutineStackFrame` interface, and the debugger and kotlinx.coroutines
+library use the interface to generate async stack traces.
+
+The next class is `ContinuationImpl`. Every continuation of a suspending function, generated by the compiler, extends this class. Note that
+the compiler does not generate restricted suspend functions (yet).
+```kotlin
+abstract class ContinuationImpl(
+    completion: Continuation<Any?>?,
+    private val _context: CoroutineContext?
+) : BaseContinuationImpl(completion) {
+    public override val context: CoroutineContext
+    protected override fun releaseIntercepted()
+
+    private var intercepted: Continuation<Any?>?
+    public fun intercepted(): Continuation<Any?>
+}
+```
+It adds the `intercepted` field and `intercepted()` function covered in the corresponding section.
+
+For restricted suspend function, there is`RestrictedContinuationImpl` class, and thus their context can only be `EmptyCoroutineContext`. It
+allows us to save several bytes when one calls `startCoroutine` on a suspending functional type, which does not inherit
+`BaseContinuationImpl`. For example, when
+the receiver is a callable reference to suspend functions
+the context of root continuation, passed to `startCoroutine` is `EmptyCoroutineContext`
+```kotlin
+abstract class RestrictedContinuationImpl(
+    completion: Continuation<Any?>?
+) : BaseContinuationImpl(completion) {
+    public override val context: CoroutineContext
+}
+```
+
+All non-restricted generated suspend lambdas extend `SuspendLambda`:
+```kotlin
+internal abstract class SuspendLambda(
+    public override val arity: Int,
+    completion: Continuation<Any?>?
+) : ContinuationImpl(completion), FunctionBase<Any?>, SuspendFunction
+```
+since all suspend lambdas are functional types, they implement the `FunctionBase` interface. `SuspendFunction` is a marker interface used in
+type checks and type conversions (see the next subsection).
+
+All restricted generated suspend lambdas extend `RestrictedSuspendLambda`:
+```kotlin
+abstract class RestrictedSuspendLambda(
+    public override val arity: Int,
+    completion: Continuation<Any?>?
+) : RestrictedContinuationImpl(completion), FunctionBase<Any?>, SuspendFunction
+```
+the only difference from `SuspendLambda` is superclass. `SuspendLambda` inherits `ContinuationImpl`, while
+`RestrictedSuspendLambda` inherits `RestrictedSuspendLambda`.
```

Comment:
This supposes to be... `RestrictedSuspendLambda` inherits `RestrictedContinuationImpl`?